GVFS/GVFS.Platform.Windows/WindowsFileBasedLock.cs

```diff
@@ -0,0 +1,135 @@
+﻿using GVFS.Common;
+using GVFS.Common.FileSystem;
+using GVFS.Common.Tracing;
+using System;
+using System.ComponentModel;
+using System.IO;
+using System.Text;
+
+namespace GVFS.Platform.Windows
+{
+    public class WindowsFileBasedLock : FileBasedLock
+    {
+        private const int HResultErrorSharingViolation = -2147024864; // -2147024864 = 0x80070020 = ERROR_SHARING_VIOLATION
+        private const int HResultErrorFileExists = -2147024816; // -2147024816 = 0x80070050 = ERROR_FILE_EXISTS
+        private const int DefaultStreamWriterBufferSize = 1024; // Copied from: http://referencesource.microsoft.com/#mscorlib/system/io/streamwriter.cs,5516ce201dc06b5f
+        private const string EtwArea = nameof(WindowsFileBasedLock);
+        private static readonly Encoding UTF8NoBOM = new UTF8Encoding(false, true); // Default encoding used by StreamWriter
+
+        private readonly object deleteOnCloseStreamLock = new object();
+        private Stream deleteOnCloseStream;
+
+        /// <summary>
+        /// FileBasedLock constructor
+        /// </summary>
+        /// <param name="lockPath">Path to lock file</param>
+        /// <param name="signature">Text to write in lock file</param>
+        /// <remarks>
+        /// GVFS keeps an exclusive write handle open to lock files that it creates with FileBasedLock.  This means that 
+        /// FileBasedLock still ensures exclusivity when the lock file is used only for coordination between multiple GVFS processes.
+        /// </remarks>
+        public WindowsFileBasedLock(
+            PhysicalFileSystem fileSystem,
+            ITracer tracer,
+            string lockPath)
+            : base(fileSystem, tracer, lockPath)
+        {
+        }
+
+        public override bool TryAcquireLock()
+        {
+            try
+            {
+                lock (this.deleteOnCloseStreamLock)
+                {
+                    if (this.deleteOnCloseStream != null)
+                    {
+                        throw new InvalidOperationException("Lock has already been acquired");

```

Comment:
@sanoursa let me know what you think of this change (to throw an exception) in my latest commit.

I made this change (rather than returning `true`) to help enforce the rule that `TryAcquireLock` should not be called on a `FileBasedLock` after it has successfully acquired the lock.

Reply:
Throwing `InvalidOperationException` makes perfect sense here. A re-entrant usage of this lock violates its contract and could result in other bugs if we don't make that obvious.